### PR TITLE
[Core] Set `linear_weights` directly on the layer

### PR DESCRIPTION
**Pull request by Yard1**
Currently, all vLLM linear layers use a `linear_weights` attribute to store a dictionary of tensors to be used for actually calculating the matmuls. Those same tensors are also set on the layer itself, leading to the tensors being referenced both in the dictionary and on the layer. If the references to the tensors on the layer are replaced (eg. by model loading code), then the tensors will stop being equal leading to incorrect matmul results. This PR removes the double reference by avoiding the storing of the `linear_weights` dictionary and instead setting the tensors directly as attributes on the layer, meaning we are left with only a single source of truth.

See https://github.com/vllm-project/vllm/pull/3476#issuecomment-2046330442 for more context

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: csrc/quantization/gptq/q_gemm.cu [hunk 1]
```diff
@@ -2067,7 +2067,7 @@ void gptq_shuffle
     const at::cuda::OptionalCUDAGuard device_guard(device_of(q_weight));
     vllm::gptq::shuffle_exllama_weight(
         (uint32_t*) q_weight.data_ptr(),
-        q_perm.device().is_meta() ? NULL : (int*) q_perm.data_ptr(),
+        q_perm.device().is_meta() || q_perm.numel() == 0 ? NULL : (int*) q_perm.data_ptr(),
```
Comment (robertgshaw2-neuralmagic) [inline csrc/quantization/gptq/q_gemm.cu:2070]: Should this file be modified?

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -29,19 +29,24 @@ class LinearMethodBase(ABC):
...
     @abstractmethod
-    def create_weights(self, input_size_per_partition: int,
+    def create_weights(self, layer: torch.nn.Module,
+                       input_size_per_partition: int,
                        output_size_per_partition: int, input_size: int,
-                       output_size: int,
-                       params_dtype: torch.dtype) -> Dict[str, Any]:
-        """Create weights for a linear layer."""
+                       output_size: int, params_dtype: torch.dtype,
+                       **extra_weight_attrs):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:36]: Could this just be an optional `weight_loader`?

I see the value of enabling `kwargs` here for future extensibility, but I don't see a case that exists yet other than `weight_loader` so far

Perhaps making the argument explicit is better until we have a reason to allow kwargs

File: vllm/model_executor/layers/quantization/gptq.py [hunk 3]
```diff
@@ -179,37 +181,40 @@ def create_weights(
...
+        layer.register_parameter("qweight", qweight)
+        set_weight_attrs(qweight, extra_weight_attrs)
+        layer.register_parameter("g_idx", g_idx)
+        set_weight_attrs(g_idx, extra_weight_attrs)
+        layer.register_parameter("qzeros", qzeros)
+        set_weight_attrs(qzeros, extra_weight_attrs)
+        layer.register_parameter("scales", scales)
+        set_weight_attrs(scales, extra_weight_attrs)
+
+        layer.exllama_state = exllama_state
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/gptq.py:194]: The `LinearLayer` does not need to know this info, so I think it should be `GPTQLinearMethod.exllama_state` ... It probably should not have been in the `weights` dict before

This will avoid having this dangling member

**Comment by robertgshaw2-neuralmagic** [comment]
LGTM, left a few minor comments

File: csrc/quantization/gptq/q_gemm.cu (hunk 1 above)
Comment (Yard1) [reply csrc/quantization/gptq/q_gemm.cu:2070]: yeah since now we need the q_perm tensor to be on device (cannot assign a meta tensor to a parameter that used to be cuda).

File: vllm/model_executor/layers/linear.py (hunk 2 above)
Comment (Yard1) [reply vllm/model_executor/layers/linear.py:36]: Actually, making it a kwarg will be easier as we don't need extra handling to _not_ set the `weight_loader` if it's left unspecified.

File: vllm/model_executor/layers/quantization/gptq.py (hunk 3 above)
Comment (Yard1) [reply vllm/model_executor/layers/quantization/gptq.py]: ah actually I don't think this is possible since GPTQLinearMethod is a singleton across all layers

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/gptq.py:194]: Sounds good

File: vllm/model_executor/layers/linear.py (hunk 2 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:36]: :)

**Review by robertgshaw2-redhat — APPROVED** [review]
